Allow menu-trigger to take a menu interface.

trshafer opened the pull request:
This should enable menus to be more easily extended for custom menu implementations.

Addresses #1560.

trshafer commented:
I went back and forth on whether or not to make it an abstract class with many of the implementation details on the abstract class. I went with the interface because it's less invasive to the consumer than an abstract class.

jelbourn commented on src/lib/menu/menu-interface.ts:
```diff
@@ -0,0 +1,11 @@
+  templateRef: TemplateRef<any>;
+  close: EventEmitter<void>;
+  _focusFirstItem: () => void;
+  _setClickCatcher: (bool: boolean) => void;
```
If these two are going to be part of this interface, they'll should lose the underscore (which implies "internal api only")

trshafer commented:
Force pushed to address the conflict in merge of #1534.

jelbourn commented on src/lib/menu/menu-interface.ts:
```diff
@@ -0,0 +1,11 @@
+import {EventEmitter, TemplateRef} from '@angular/core';
+import {MenuPositionX, MenuPositionY} from './menu-positions';
+
+export interface MdMenuInterface {
```
Let's call this `MdMenuPanel`
(generally trying to avoid using the word "interface" in the names of interfaces)

trshafer commented:
Hey @jelbourn, I have made the requested updates. Good suggestions. The test really helps demonstrate the creation of custom menus.

jelbourn commented:
LGTM aside from a couple last comments. @kara should also take a look

kara commented on src/lib/menu/menu-panel.ts:
```diff
@@ -0,0 +1,10 @@
+  positionY: MenuPositionY;
+  templateRef: TemplateRef<any>;
+  close: EventEmitter<void>;
+  _focusFirstItem: () => void;
```
Agree with @jelbourn that underscore should be removed if possible

trshafer commented:
Hey @kara, thanks for the review. I have removed the underscore from the interface, rebased, squashed my commits, and updated the commit message style.
